components/ContainedInputViews/src/MDCContainedInputView.h

```diff
@@ -0,0 +1,319 @@
+ */
+@property(nonatomic, assign) CGFloat customUnderlineLabelDrawPriority;
+
+/**
+ When set to YES, the floating label floats when the view becomes the first responder. When
+ set to NO it disappears.
+
+ @note The default is YES.
+ */
+@property(nonatomic, assign) BOOL canFloatingLabelFloat;
+
+/**
+ This property toggles the error state (similar to @c isHighlighted, @c isEnabled, @c isSelected,
+ etc.) that is part of a general interpretation of the states outlined in the Material guidelines
+ for Text Fields. See the @c MDCContainedInputViewState enum for more information.
+ */
+@property(nonatomic, assign) BOOL isErrored;
+
+/**
+ This property toggles the activated state (similar to @c isHighlighted, @c isEnabled, @c
+ isSelected, etc.) that is part of a general interpretation of the states outlined in the Material
+ guidelines for Text Fields. See the @c MDCContainedInputViewState enum for more information.
+ */
+@property(nonatomic, assign) BOOL isActivated;
+
+/**
+  This method returns a color scheme for a given state.
+ */
+- (nonnull id<MDCContainedInputViewColorScheming>)containedInputViewColorSchemingForState:
+    (MDCContainedInputViewState)containedInputViewState;
+
+/**
+ This method sets a color scheme for a given state.
+ */
+- (void)setContainedInputViewColorScheming:
+            (nonnull id<MDCContainedInputViewColorScheming>)containedInputViewColorScheming
+                                  forState:(MDCContainedInputViewState)textFieldState;
+
+/**
+ Returns the rect surrounding the main content, i.e. the area that the container should be drawn
+ around.
+ */
+@property(nonatomic, assign, readonly) CGRect containerFrame;
+
+/**
+ This API allows the user to override the default main content area height. The main content area is
+ the part of the view where the where the data input happens. It is located above the underline
+ label area. If this property is set to a value that's lower than the default main content area
+ height the value will be ignored in the calculation of the view's @c intrinsicContentSize.
+ */
+@property(nonatomic, assign) CGFloat preferredMainContentAreaHeight;
+
+/**
+ This API allows the user to override the default underline label area height. The underline label
+ area is the part of the view where the underline labels are. It is located below the main content
+ area. If this property is set to a value that's lower than the default underline label area height
+ the value will be ignored in the calculation of the view's @c intrinsicContentSize.
+ */
+@property(nonatomic, assign) CGFloat preferredUnderlineLabelAreaHeight;
+
+@end
+
+/**
+ This protocol represents a set of colors that are semantically meaningful and specific to
+ MDCContainedInputView. Each property corresponds to the color of one or more views that an
+ MDCContainedInputView manages at a given point of time.
+ */
+@protocol MDCContainedInputViewColorScheming <NSObject>
```

Comment:
This name is misleading. Color scheming is used by the color subsystem to group semantic color names. This protocol groups attributes which happen to be colors. Note that Material state theming also use elevation, which is not a color, and may use more visual indicators in the future which many not be colors.

Reply:
I wouldn't say it just _happens_ to have colors. The fact that it has colors and only colors is intentional. Maybe we could change it to have other state dependent information as well? I'm not sure what other state dependent information there is. You keep mentioning elevation, but I don't know of any elevation stuff that's relevant here. I've never heard elevation discussed in the context of text fields, which I'm using as my reference for all this stuff. I could be wrong though!

Reply:
MDCContainedInputView provides the state dependent accessors you're describing. Better yet, the theming extension (not a themer!) uses it to achieve theming.